plumbing/msg/waiter.go

```diff
@@ -175,9 +177,8 @@ func (w *Waiter) waitForMessage(ctx context.Context, ch <-chan interface{}, msgC
 func (w *Waiter) receiptFromTipSet(ctx context.Context, msgCid cid.Cid, ts types.TipSet) (*types.MessageReceipt, error) {
 	// Receipts always match block if tipset has only 1 member.
 	var rcpt *types.MessageReceipt
-	blks := ts.ToSlice()
-	if len(ts) == 1 {
-		b := blks[0]
+	if ts.IsSolo() {
```

Comment:
Perhaps now that we have a tipset struct, some future work would be cache receipts in tip sets. That would keep us from having to run a full state transition to fetch a single receipt when there's more than one block. This method is pretty ugly.

Reply:
This is probably the ugliest code in the entire project and it should be made better with or without caching receipts.  That said caching receipts is a great idea.  We should probably think about this alongside caching aggregate state cids / aggregate state trees, and limiting the in-memory tipset cache.